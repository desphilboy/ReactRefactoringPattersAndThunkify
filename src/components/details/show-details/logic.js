import { resolveType } from '../../../redux/effects/type.js';
import { isValidCache } from '../../../redux/helpers.js';

const nameTypesSelector = state => state.lookUp.nameTypes;
const tipSelector = state => state.lookUp.inProgress;

export const mapStateToProps = state => ({
	nameTypes: nameTypesSelector(state),
	typeInProgress: tipSelector(state),
});

export const mapDispatchToProps = dispatch => ({
	getType: petName => dispatch(resolveType({ petName })),
});

export const extraProps = ({ petName, nameTypes }) => ({
	nameType: nameTypes.find(nt => nt.name === petName),
});

export const callTypeApi = ({ getType, petName }) => {
	getType(petName);
};

export const callTypeInProgress = ({ typeInProgress }) => !!typeInProgress;

export const callTypeApiNeeded = ({ nameType }) => !nameType || !isValidCache(nameType);
