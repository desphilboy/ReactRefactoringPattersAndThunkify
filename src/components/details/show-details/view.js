import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { ClockLoader } from 'react-spinners';
import { override } from '../styles.js';
import InfoComponent from '../show-info';

export const TopSpinner = () => (
	<div className="sweet-loading">
		<ClockLoader css={override} size={150} color="#123abc" loading />
	</div>
);

export const ShowDetailsBody = ({ petName, nameType: { petType } = {} }) => (
	<div>
		<h2>
			{petName} is a {petType}
		</h2>
		<InfoComponent petType={petType} />
	</div>
);
