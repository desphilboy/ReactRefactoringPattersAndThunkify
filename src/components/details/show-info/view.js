import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { ClipLoader } from 'react-spinners';
import { resolveInfo } from '../../../redux/effects/info.js';
import { override } from '../styles.js';
import withInit from '../../higher-order/with-init';
import withProps from '../../higher-order/with-props';
import { isValidCache } from '../../../redux/helpers.js';

const BottomSpinner = () => (
	<div className="sweet-loading">
		<ClipLoader css={override} size={150} color="#123abc" loading />
	</div>
);

const ShowInfoView = ({ info: { len, width, height, weight, eats, legs_hands, description } }) => (
	<div>
		<p>
			dimensions: L:{len} W:{width} H:{height}
		</p>
		<p>eats: {(eats || []).join(',')}</p>
		<p>body shape: has {legs_hands} hands and legs</p>
		<p>
			Weight {weight && weight.min} to {weight && weight.max} Kgs.
		</p>
		<p>{description}</p>
	</div>
);

const findTypeInfo = (info, petType) => info.find(inf => inf.petType === petType);
const animalsInfoSelector = state => state.info.animals;
const iipSelector = state => state.info.inProgress;

const mapStateToProps = state => ({
	infoInProgress: iipSelector(state),
	animals: animalsInfoSelector(state),
});

const mapDispatchToProps = dispatch => ({
	getInfo: type => dispatch(resolveInfo(type)),
});

const extraProps = ({ animals, petType }) => ({
	info: findTypeInfo(animals, petType),
});

const callInfoApi = ({ getInfo, petType }) => {
	getInfo(petType);
};

const callInfoInProgress = ({ infoInProgress }) => infoInProgress;

const callInfoNeeded = ({ info }) => !isValidCache(info);

export default compose(
	connect(mapStateToProps, mapDispatchToProps),
	withProps(extraProps),
	withInit(callInfoApi, callInfoNeeded, callInfoInProgress, BottomSpinner),
)(ShowInfoView);
