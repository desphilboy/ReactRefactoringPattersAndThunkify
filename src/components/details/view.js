import React from "react";
import { connect } from "react-redux";
import { Field, reduxForm } from "redux-form";
import { useHistory } from "react-router-dom";
import { css } from "@emotion/core";
import { ClipLoader, ClockLoader } from "react-spinners";
import {
	setPetName,
	setPetType
} from "../../redux/reducers/pageInfo.reducer.js";
import { resolveType } from "../../redux/effects/type.js";
import { resolveInfo } from "../../redux/effects/info.js";

const override = css`
	display: block;
	margin: 0 auto;
	border-color: red;
`;

class Details extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			typeResolved: false,
			infoResolved: false
		};
	}

	componentDidMount() {
		this.props.getType(this.props.petName);
	}

	componentDidUpdate(prevProps) {
		console.log(prevProps.typeInProgress, this.props.typeInProgress);
		if (prevProps.typeInProgress && !this.props.typeInProgress) {
			console.log("setting the state");
			this.setState({ typeResolved: true });
			console.log(this.state);
			setTimeout(() => this.props.getInfo(this.props.petType), 200);
		}
		if (prevProps.infoInProgress && !this.props.infoInProgress) {
			console.log("setting the state");
			this.setState({ infoResolved: true });

			console.log(this.state);
		}
	}

	render(props) {
		return (
			<div>
				<h1>Hello dear owner of {this.props.petName}</h1>

				{!this.state.typeResolved && (
					<div className="sweet-loading">
						<ClockLoader css={override} size={150} color="#123abc" loading />
					</div>
				)}
				{this.state.typeResolved && (
					<h2>
						{this.props.petName} is a {this.props.petType}
					</h2>
				)}
				{!this.state.infoResolved && (
					<div className="sweet-loading">
						<ClipLoader css={override} size={150} color="#123abc" loading />
					</div>
				)}
				{this.state.infoResolved && (
					<h2>
						{(() => {
							const info = this.props.info.find(
								inf => inf.petType === this.props.petType
							);
							return (
								<div>
									<p>
										dimensions: L:
										{info.len} W:
										{info.width} H:
										{info.height}
									</p>
									<p>eats: {info.eats.join(",")}</p>
									<p>body shape: has {info.legs_hands} hands and legs</p>
									<p>
										Weight {info.weight.min} to {info.weight.max} Kgs.
									</p>
									<p>{info.description} </p>
								</div>
							);
						})()}
					</h2>
				)}
			</div>
		);
	}
}

export default connect(
	state => ({
		petName: state.pageInfo.petName,
		petType: state.pageInfo.petType,
		typeInProgress: state.lookUp.inProgress,
		infoInProgress: state.info.inProgress,
		info: state.info.animals
	}),
	dispatch => ({
		getType: name => dispatch(resolveType(name)),
		getInfo: type => dispatch(resolveInfo(type))
	})
)(Details);
