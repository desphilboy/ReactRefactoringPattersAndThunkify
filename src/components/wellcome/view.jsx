import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Field, reduxForm } from 'redux-form';
import { setPetName } from '../../redux/reducers/pageInfo.reducer.js';
import withProps from '../higher-order/with-props';

const onSubmit = ({ history, setPetName }) => ({ typedName, selectedName }) => {
  console.log('History>>>>>>>>', history);
  console.log(typedName, selectedName);
  const name = typedName || selectedName;
  setPetName(name);
  history.push('/profile');
};

const nameTypesSelector = state => state.lookUp.nameTypes;

const mapNameTypesToList = nameTypes => nameTypes.map(nt => nt.name);

const extraProps = ({ setPetName, history }) => ({
  onSubmit: onSubmit({ history, setPetName }),
});

const mapStateToProps = state => ({
  petsList: mapNameTypesToList(nameTypesSelector(state)),
});

const mapDispatchToProps = dispatch => ({
  setPetName: name => dispatch(setPetName(name)),
});

export const WellComeView = ({ petsList, handleSubmit, onSubmit }) => (
  <div>
    <h1> Wellcome to best FEnders! FrEnds! website</h1>
    <h2>Hi dear beloved owner</h2>
    <form onSubmit={handleSubmit(onSubmit)}>
      {!!petsList.length && (
        <div>
          please select or Input pets name:
          <Field key="1" name="selectedName" component="select" width="20">
            <option key="noValue" />
            {petsList.map((pet, index) => (
              <option key={index} value={pet}>
                {pet}
              </option>
            ))}
          </Field>
          <br />
          or:
          <br />
        </div>
      )}
      <label>Input your pet's name: </label>
      <Field key="2" name="typedName" component="input" type="text" />
      <br />
      <button style={{ margin: '50px' }} type="submit">
        Submit
      </button>
    </form>
  </div>
);

export default compose(
  withRouter,
  connect(mapStateToProps, mapDispatchToProps),
  withProps(extraProps),
  reduxForm({ form: 'petNameForm' }),
)(WellComeView);
