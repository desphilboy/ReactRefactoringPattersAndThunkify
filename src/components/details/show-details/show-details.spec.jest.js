import React from 'react';
import component from './index.js';
import { shallow } from 'enzyme';
import { mapStateToProps, mapDispatchToProps, extraProps } from './logic.js';
import { TopSpinner, ShowDetailsBody } from './view.js';

describe('ShowDetails', () => {
	describe('index', () => {
		describe('component', () => {
			it('matches the snapshot', () => {
				expect(component.displayName).toMatchInlineSnapshot(`"Connect(withProps(withInit(ShowDetailsBody)))"`);
			});
		});
	});

	describe('logic', () => {
		describe('mapStateToProps', () => {
			let map;
			beforeEach(() => {
				map = mapStateToProps(sampleState);
			});

			it('brings nameTypes from the state', () => {
				expect(map.nameTypes).toEqual([{ name: 'n', petType: 'pt' }]);
			});

			it('brings inProgress from state lookup', () => {
				expect(map.typeInProgress).toEqual('whatever');
			});
		});

		describe('extraProps', () => {
			let map;
			beforeEach(() => {
				map = extraProps({
					nameTypes: [
						{ name: 'n', petType: 'pt' },
						{ name: 'nn', petType: 'pptt' },
					],
					petName: 'n',
				});
			});

			it('brings nameTypes from the state', () => {
				expect(map.nameType).toEqual({ name: 'n', petType: 'pt' });
			});

			it('if name not found returns undefined', () => {
				map = extraProps({ nameTypes: [{ name: 'n', petType: 'pt' }], petName: 't' });
				expect(map.nameType).toBeUndefined();
			});
		});
	});

	describe('views', () => {
		describe('TopSpinner', () => {
			it('match the snapshot', () => {
				const wrapper = shallow(<TopSpinner />);
				expect(wrapper).toMatchSnapshot();
			});
		});

		describe('ShowDetailsBody', () => {
			it('match the snapshot with invalid params', () => {
				const wrapper = shallow(<ShowDetailsBody />);
				expect(wrapper).toMatchSnapshot();
			});

			it('match the snapshot with params', () => {
				const wrapper = shallow(<ShowDetailsBody petName="maloo" nameType={{ maloo: 'baloo', petType: 'paloo' }} />);
				expect(wrapper).toMatchSnapshot();
			});
		});
	});
});

const sampleState = { lookUp: { nameTypes: [{ name: 'n', petType: 'pt' }], inProgress: 'whatever' } };
