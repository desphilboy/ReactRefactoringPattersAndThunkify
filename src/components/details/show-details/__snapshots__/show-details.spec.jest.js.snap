// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowDetails views ShowDetailsBody match the snapshot with invalid params 1`] = `ShallowWrapper {}`;

exports[`ShowDetails views ShowDetailsBody match the snapshot with params 1`] = `ShallowWrapper {}`;

exports[`ShowDetails views TopSpinner match the snapshot 1`] = `ShallowWrapper {}`;
